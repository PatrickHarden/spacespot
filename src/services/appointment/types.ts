export interface AppointmentState {
  data?: Array<Appointment>
  error?: string
}

export interface Appointment {
  id: number
  enquireId: number
  description: string
  peer1Id: string
  peer2Id: string
  proposedTime: string
  createdAt: string
  lastUpdatedAt: string
  lastProposedBy: string
  appointmentStatus: string
}

export interface NewAppointment {
  description: string
  proposedTime: string
}
export interface EditAppointment {
  id: string
  description: string
  proposedTime: string
}
export interface AppointmentAction {
  type: string
  payload?: Array<Appointment> | NewAppointment | EditAppointment | string
}
