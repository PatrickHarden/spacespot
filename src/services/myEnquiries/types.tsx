import { TermStatus, NegotiationTerm } from 'services/negotiation/types'
import { ChatMessage } from 'services/chat/types'
import { EnquiryType } from 'services/enquiry/types'

export type EventType =
  | 'NEW_APPOINTMENT'
  | 'MODIFIED_APPOINTMENT'
  | 'ACCEPT_APPOINTMENT'
  | 'CANCEL_APPOINTMENT'
  | 'DOCUMENT_UPLOAD'
  | 'DOCUMENT_DELETE'
  | 'COUNTER_NEGOTIATION'
  | 'ACCEPT_NEGOTIATION'
  | 'REJECT_NEGOTIATION'
  | 'SPECIAL_PROVISION_NEGOTIATION'
  | 'ACCEPT_SPECIAL_PROVISION'
  | 'REJECT_SPECIAL_PROVISION'
  | 'LEASE_SIGNED'
  | 'ACCEPT_LEASE'

/**
 *  Response from /myenquiries
 */
export interface MyEnquiriesAPIResponse {
  enquireId: number
  negotiationId: number
  enquireStatus: TermStatus
  spaceId: string
  terms: Array<NegotiationTerm>
  messages: Array<ChatMessage>
  lastReadByTenant?: string
  lastReadByLandlord?: string
  flex?: boolean
  flexType?: EnquiryType
  flexIdentifier?: number
  /* Tenant */
  peer1Id: string
  /* Landlord */
  peer2Id: string
  otherPeerDetails: {
    displayName?: string
  }
  lastActivityDetails?: {
    eventType: EventType
    eventDate: string
  }
  eventNotification?: boolean
}

export interface MyEnquiriesState {
  data?: Array<MyEnquiriesAPIResponse>
  loading: boolean
  error?: string
  notifications?: number
}
