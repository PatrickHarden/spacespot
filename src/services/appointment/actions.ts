import {
  AppointmentAction,
  Appointment,
  EditAppointment,
  NewAppointment,
} from './types'
import constants from './constants'

const get = (): AppointmentAction => ({
  type: constants.GET,
})
const getSuccess = (payload: Array<Appointment>) => ({
  type: constants.GET_SUCCESS,
  payload,
})
const getError = (payload: string): AppointmentAction => ({
  type: constants.GET_ERROR,
  payload,
})
const update = (): AppointmentAction => ({
  type: constants.UPDATE,
})

const create = (payload: NewAppointment): AppointmentAction => ({
  type: constants.CREATE,
  payload,
})

const createSuccess = (payload: Appointment) => ({
  type: constants.CREATE_SUCCESS,
  payload,
})
const createError = (payload: string): AppointmentAction => ({
  type: constants.CREATE_ERROR,
  payload,
})

const accept = (payload: string): AppointmentAction => ({
  type: constants.ACCEPT,
  payload,
})

const acceptSuccess = (payload: Appointment) => ({
  type: constants.ACCEPT_SUCCESS,
  payload,
})

const acceptError = (payload: string): AppointmentAction => ({
  type: constants.ACCEPT_ERROR,
  payload,
})

const cancel = (payload: string): AppointmentAction => ({
  type: constants.CANCEL,
  payload,
})

const cancelSuccess = (payload: Appointment) => ({
  type: constants.CANCEL_SUCCESS,
  payload,
})

const cancelError = (payload: string): AppointmentAction => ({
  type: constants.CANCEL_ERROR,
  payload,
})

const change = (payload: EditAppointment): AppointmentAction => ({
  type: constants.CHANGE,
  payload,
})

const changeSuccess = (payload: Appointment) => ({
  type: constants.CHANGE_SUCCESS,
  payload,
})

const changeError = (payload: string): AppointmentAction => ({
  type: constants.CHANGE_ERROR,
  payload,
})
export default {
  get,
  getSuccess,
  getError,
  update,
  create,
  createSuccess,
  createError,
  accept,
  acceptSuccess,
  acceptError,
  cancel,
  cancelSuccess,
  cancelError,
  change,
  changeSuccess,
  changeError,
}
