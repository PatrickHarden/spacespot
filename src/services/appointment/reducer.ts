import constants from './constants'
import { AppointmentAction, AppointmentState, Appointment } from './types'

const initialState: AppointmentState = {}

const reducer = (
  state: AppointmentState = initialState,
  action: AppointmentAction,
): AppointmentState => {
  switch (action.type) {
    case constants.GET_ERROR:
      return { ...state, error: action.payload as string }
    case constants.GET_SUCCESS:
      return { ...state, data: action.payload as Array<Appointment> }
    case constants.CREATE_SUCCESS:
      return {
        ...state,
        data: [...(state.data || []), action.payload as Appointment],
      }
    default:
      return state
  }
}
export default reducer
