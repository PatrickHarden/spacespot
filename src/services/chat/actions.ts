import { ChatAction, ChatResponse, ChatMessage } from './types'
import constants from './constants'

const get = (): ChatAction => ({
  type: constants.GET,
})
const getSuccess = (payload: ChatResponse) => ({
  type: constants.GET_SUCCESS,
  payload,
})
const getError = (payload: string): ChatAction => ({
  type: constants.GET_ERROR,
  payload,
})

const newMessage = (payload: string) => ({
  type: constants.NEW_MESSAGE,
  payload,
})
const newMessageSuccess = (payload: ChatMessage) => ({
  type: constants.NEW_MESSAGE_SUCCESS,
  payload,
})
const newMessageError = (payload: string) => ({
  type: constants.NEW_MESSAGE_ERROR,
  payload,
})
export default {
  get,
  getSuccess,
  getError,
  newMessage,
  newMessageSuccess,
  newMessageError,
}
