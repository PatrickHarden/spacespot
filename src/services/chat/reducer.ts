import { get } from 'lodash'
import constants from './constants'
import { ChatAction, ChatState, ChatResponse, ChatMessage } from './types'

const initialState: ChatState = {}
const addNewMessage = (
  data: ChatResponse,
  payload: ChatMessage,
): ChatResponse => {
  const messagesOld: Array<ChatMessage> = get(data, 'messages')
  return { ...data, messages: [...messagesOld, payload] }
}
const reducer = (
  state: ChatState = initialState,
  action: ChatAction,
): ChatState => {
  switch (action.type) {
    case constants.GET:
      return { ...state }
    case constants.GET_ERROR:
      return { ...state, error: action.payload as string }
    case constants.GET_SUCCESS:
      return { ...state, data: action.payload as ChatResponse }
    case constants.NEW_MESSAGE_SUCCESS:
      return {
        ...state,
        data: addNewMessage(
          state.data as ChatResponse,
          action.payload as ChatMessage,
        ),
      }
    default:
      return state
  }
}
export default reducer
