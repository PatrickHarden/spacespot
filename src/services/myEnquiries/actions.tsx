import { MyEnquiriesAPIResponse } from './types'
import constants from './constants'

const get = () => ({
  type: constants.GET,
})

const getSuccess = (payload: Array<MyEnquiriesAPIResponse>) => ({
  type: constants.GET_SUCCESS,
  payload,
})

const getError = (payload: string) => ({
  type: constants.GET_ERROR,
  payload,
})
const accept = (payload: string) => ({
  type: constants.ACCEPT,
  payload,
})

const acceptSuccess = (payload: MyEnquiriesAPIResponse) => ({
  type: constants.ACCEPT_SUCCESS,
  payload,
})

const acceptError = (payload: string) => ({
  type: constants.ACCEPT_ERROR,
  payload,
})
const reject = (payload: string) => ({
  type: constants.REJECT,
  payload,
})

const rejectSuccess = (payload: MyEnquiriesAPIResponse) => ({
  type: constants.REJECT_SUCCESS,
  payload,
})

const rejectError = (payload: string) => ({
  type: constants.REJECT_ERROR,
  payload,
})

const cancel = (payload: { enquiryId: string; feedback?: string }) => ({
  type: constants.CANCEL,
  payload,
})

const cancelSuccess = (payload: MyEnquiriesAPIResponse) => ({
  type: constants.CANCEL_SUCCESS,
  payload,
})

const cancelError = (payload: string) => ({
  type: constants.CANCEL_ERROR,
  payload,
})

const setNotifications = (payload: number) => ({
  type: constants.SET_NOTIFICATIONS,
  payload,
})

const getNotifications = () => ({
  type: constants.GET_NOTIFICATIONS,
})

export default {
  get,
  getSuccess,
  getError,
  accept,
  acceptSuccess,
  acceptError,
  reject,
  rejectSuccess,
  rejectError,
  cancel,
  cancelSuccess,
  cancelError,
  setNotifications,
  getNotifications,
}
