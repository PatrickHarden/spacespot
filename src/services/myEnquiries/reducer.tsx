import constants from './constants'
import { MyEnquiriesState, MyEnquiriesAPIResponse } from './types'
import { TermStatus } from 'services/negotiation/types'

const initialState: MyEnquiriesState = { data: undefined, loading: false }

const getUpdatedData = (
  payload: MyEnquiriesAPIResponse,
  data: Array<MyEnquiriesAPIResponse>,
  termStatus: TermStatus,
) =>
  data.map((enquiry: any) =>
    enquiry.enquireId === payload.enquireId
      ? { ...enquiry, enquireStatus: termStatus }
      : enquiry,
  )

const TranslatingActionToTerms = {
  [constants.ACCEPT_SUCCESS]: TermStatus.Accepted,
  [constants.REJECT_SUCCESS]: TermStatus.Declined,
  [constants.CANCEL_SUCCESS]: TermStatus.Closed,
}

const reducer = (state = initialState, action: any) => {
  switch (action.type) {
    case constants.GET:
      return { ...state, loading: true }
    case constants.GET_ERROR:
      return { error: action.payload as string, loading: false }
    case constants.GET_SUCCESS:
      return {
        ...state,
        data: action.payload,
        loading: false,
      }
    case constants.ACCEPT_SUCCESS:
    case constants.REJECT_SUCCESS:
    case constants.CANCEL_SUCCESS:
      return {
        ...state,
        data: getUpdatedData(
          action.payload,
          state.data as Array<MyEnquiriesAPIResponse>,
          TranslatingActionToTerms[action.type],
        ),
      }
    case constants.SET_NOTIFICATIONS:
      return {
        ...state,
        notifications: action.payload,
      }
    default:
      return state
  }
}

export default reducer
