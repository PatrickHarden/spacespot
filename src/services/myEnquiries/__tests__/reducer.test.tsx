import reducer from '../reducer'
import actions from '../actions'
import { MyEnquiriesAPIResponse, MyEnquiriesState } from '../types'

import data1 from '../__mocks__/user-enquiries.json'

describe('MyEnquiries reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual({
      data: undefined,
      loading: false,
    })
  })
  it('init should update state', () => {
    const resp = { loading: true }
    expect(reducer(undefined, actions.get())).toEqual(resp)
  })

  it('error should update state', () => {
    const err = { error: 'Error1' }
    expect(reducer(undefined, actions.getError('Error1'))).toEqual({
      ...err,
      loading: false,
    })
  })

  it('get success should update state', () => {
    const resp = {} as MyEnquiriesAPIResponse
    expect(reducer(undefined, actions.getSuccess(resp))).toEqual({
      data: resp,
      loading: false,
    })
  })

  it('get accept and reject success should update state', () => {
    const stateMyEnquiries1: MyEnquiriesState = {
      data: data1 as Array<MyEnquiriesAPIResponse>,
      loading: false,
    }
    const respAccept = {
      enquireId: 167,
      negotiationId: 2,
    } as MyEnquiriesAPIResponse
    expect(
      reducer(stateMyEnquiries1, actions.acceptSuccess(respAccept)),
    ).toMatchSnapshot()
    const respReject = {
      enquireId: 168,
      negotiationId: 2,
    } as MyEnquiriesAPIResponse
    expect(
      reducer(stateMyEnquiries1, actions.rejectSuccess(respReject)),
    ).toMatchSnapshot()
  })
})
