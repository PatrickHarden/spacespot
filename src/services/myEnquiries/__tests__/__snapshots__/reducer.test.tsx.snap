// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyEnquiries reducer get accept and reject success should update state 1`] = `
Object {
  "data": Array [
    Object {
      "enquireId": 167,
      "enquireStatus": "ACCEPTED",
      "messages": Array [
        Object {
          "message": "",
          "sentAt": "2020-02-25T16:59:53.122+0000",
          "userId": "de2563f1-2e15-4a6a-9e39-9600a8ad4ec1",
        },
      ],
      "negotiationId": 147,
      "spaceId": "SS_SPACE_650",
      "terms": Array [
        Object {
          "termId": 1,
          "value": "12",
        },
        Object {
          "termId": 2,
          "value": "12.0",
        },
        Object {
          "termId": 3,
          "value": "Tue Feb 25 2020 17:59:47 GMT+0100 (hora est치ndar de Europa central)",
        },
        Object {
          "termId": 4,
          "value": "123",
        },
        Object {
          "termId": 5,
          "value": "",
        },
        Object {
          "termId": 6,
          "value": "",
        },
        Object {
          "termId": 7,
          "value": "",
        },
        Object {
          "termId": 8,
          "value": "123.0",
        },
        Object {
          "termId": 9,
          "value": "",
        },
      ],
    },
    Object {
      "enquireId": 168,
      "enquireStatus": "PENDING_TO_ACCEPT",
      "messages": Array [
        Object {
          "message": "",
          "sentAt": "2020-02-25T17:00:34.014+0000",
          "userId": "de2563f1-2e15-4a6a-9e39-9600a8ad4ec1",
        },
      ],
      "negotiationId": 148,
      "spaceId": "SS_SPACE_650",
      "terms": Array [
        Object {
          "termId": 1,
          "value": "12",
        },
        Object {
          "termId": 2,
          "value": "12.0",
        },
        Object {
          "termId": 3,
          "value": "Tue Feb 25 2020 18:00:28 GMT+0100 (hora est치ndar de Europa central)",
        },
        Object {
          "termId": 4,
          "value": "13",
        },
        Object {
          "termId": 5,
          "value": "",
        },
        Object {
          "termId": 6,
          "value": "",
        },
        Object {
          "termId": 7,
          "value": "",
        },
        Object {
          "termId": 8,
          "value": "123.0",
        },
        Object {
          "termId": 9,
          "value": "",
        },
      ],
    },
  ],
  "loading": false,
}
`;

exports[`MyEnquiries reducer get accept and reject success should update state 2`] = `
Object {
  "data": Array [
    Object {
      "enquireId": 167,
      "enquireStatus": "PENDING_TO_ACCEPT",
      "messages": Array [
        Object {
          "message": "",
          "sentAt": "2020-02-25T16:59:53.122+0000",
          "userId": "de2563f1-2e15-4a6a-9e39-9600a8ad4ec1",
        },
      ],
      "negotiationId": 147,
      "spaceId": "SS_SPACE_650",
      "terms": Array [
        Object {
          "termId": 1,
          "value": "12",
        },
        Object {
          "termId": 2,
          "value": "12.0",
        },
        Object {
          "termId": 3,
          "value": "Tue Feb 25 2020 17:59:47 GMT+0100 (hora est치ndar de Europa central)",
        },
        Object {
          "termId": 4,
          "value": "123",
        },
        Object {
          "termId": 5,
          "value": "",
        },
        Object {
          "termId": 6,
          "value": "",
        },
        Object {
          "termId": 7,
          "value": "",
        },
        Object {
          "termId": 8,
          "value": "123.0",
        },
        Object {
          "termId": 9,
          "value": "",
        },
      ],
    },
    Object {
      "enquireId": 168,
      "enquireStatus": "REJECTED",
      "messages": Array [
        Object {
          "message": "",
          "sentAt": "2020-02-25T17:00:34.014+0000",
          "userId": "de2563f1-2e15-4a6a-9e39-9600a8ad4ec1",
        },
      ],
      "negotiationId": 148,
      "spaceId": "SS_SPACE_650",
      "terms": Array [
        Object {
          "termId": 1,
          "value": "12",
        },
        Object {
          "termId": 2,
          "value": "12.0",
        },
        Object {
          "termId": 3,
          "value": "Tue Feb 25 2020 18:00:28 GMT+0100 (hora est치ndar de Europa central)",
        },
        Object {
          "termId": 4,
          "value": "13",
        },
        Object {
          "termId": 5,
          "value": "",
        },
        Object {
          "termId": 6,
          "value": "",
        },
        Object {
          "termId": 7,
          "value": "",
        },
        Object {
          "termId": 8,
          "value": "123.0",
        },
        Object {
          "termId": 9,
          "value": "",
        },
      ],
    },
  ],
  "loading": false,
}
`;
